mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -932,10 +940,26 @@ LogicalResult Parser::parseDirective(SmallVectorImpl<ast::Decl *> &decls) {
   StringRef directive = curToken.getSpelling();
   if (directive == "#include")
     return parseInclude(decls);
+  if (directive == "#once")
+    return parseOnce(decls);
 
   return emitError("unknown directive `" + directive + "`");
 }
 
+LogicalResult Parser::parseOnce(SmallVectorImpl<ast::Decl *> &decls) {
+  consumeToken(Token::directive);
+
+  // #once on main file has no effect
```

Comment:
Why have this special case here, instead of adding it also to the `alreadyIncludedFiles` set?

Reply:
Neither the Lexer nor the SourceMgr know about the path of the main file. This would require some refactoring to pass the string through different APIs. I think current solution is simpler and still make sense since you would not expect to see the main file in the include stack storage